DAS-2295: Adds example notebook

flamingbear opened the pull request:
## Description

- Adds example notebook to demonstrate the service and examine the output data.

## Jira Issue ID

[DAS-2295](https://bugs.earthdata.nasa.gov/browse/DAS-2295)

## Local Test Steps

Create python environment with the requirements in docs/docs-requirements.txt

Activate the environment and run the notebook.




## PR Acceptance Checklist
* [] Jira ticket acceptance criteria met.
* [X] `CHANGELOG.md` updated to include high level summary of PR changes.
* [N/A] `docker/service_version.txt` updated if publishing a release.
* [N/A] Tests added/updated and passing.
* [X] Documentation updated (if needed).

flamingbear committed:
DAS-2295: Adds example notebook

updates README

flamingbear committed:
DAS-2295: tweaking notebook.

flamingbear committed:
DAS-2295: store notebook rendered?

flamingbear committed:
DAS-2295: swap strings

flamingbear committed:
DAS-2295: update readme and changelog

flamingbear commented on docs/example-usage.ipynb:
I added the "to follow along" as a manual step because I didn't know how to just wget or curl the file with redirects and authentication.

owenlittlejohns replied:
I think that is a good way to go.

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+   "source": [
+    "# Harmony SMAP L2 gridding service\n",
+    "\n",
+    "This notebook describes the basic functionality of the service and how to call from `harmony-py`"
```
I think something that would be valuable is a sentence right at the top that describes exactly what it is this service does (at the 10,000 foot level). The service name is fairly descriptive and basically covers it, but I'm just thinking something along the lines of:

> This service takes swath data products from the SMAP mission as an input and uses row and column information from within the file to place the data for each variable on to an Equal-Area Scalable Earth (EASE) Grid 2.0.

I think the main value here is in letting someone know, straight at the top, "is this service for me?"

flamingbear replied:
151e797

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+    "| C1268429748-EEDTEST | SPL2SMAP   | SMAP L2 Radar/Radiometer Half-Orbit 9 km EASE-Grid Soil Moisture    |\n",
+    "| C1268429309-EEDTEST | SPL2SMP    | SMAP L2 Radiometer Half-Orbit 36 km EASE-Grid Soil Moisture         |\n",
+    "\n",
+    "Any of these collections can be used with the smap-l2-gridder."
```
Total nit: I think it would be better to use "Harmony SMAP L2 Gridding Service" here instead of "smap-l2-gridder". Aside from increased consistency, this currently feels like we're leaking an internal name that is used as shorthand in configuration, etc.

flamingbear replied:
151e797

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+   "source": [
+    "### What is L2 Gridded data?\n",
+    "\n",
+    "Level 2 Gridded Data is data that is stored in 1D arrays along with row and column indices.\n",
```
Another nit: Might be worth expanding on "row and column indices"... for what? Maybe:

> Level 2 Gridded data stores swath data in 1-D arrays, with corresponding 1-D arrays storing EASE Grid-2.0 column and row indices for each data point.

flamingbear replied:
151e797

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
```
Nit: The notebook has been committed in an executed state, which it probably shouldn't be.

flamingbear replied:
I thought I mentioned I did that on purpose so that you could see it in the repo without having to execute it.

owenlittlejohns replied:
Discussed in Slack - leaving as-is, where there are no leaked credentials in the notebook JSON and there is value in the outputs being pre-rendered, as @flamingbear says above.

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+   "source": [
+    "## Using the service\n",
+    "\n",
+    "use [harmony-py](https://github.com/nasa/harmony-py) and initialize a UAT Harmony client"
```
Super duper nit: There's some missing capitalisation in this cell:

* "use harmony-py..." -> "Use harmony-py...".
* "a crs of WGS84..." -> "a CRS of WGS84..."

I'm only mentioning this level of nitpick as this is end-user documentation.

owenlittlejohns replied:
Also, I'm not sure if you need a tiny bit more context for what harmony-py is. That is, maybe just saying it is a Python-based client for making requests against Harmony.

Feel free to ignore this second bit, I guess I'm trying to be cautious in assuming how much prior knowledge of the Harmony ecosystem someone reading this notebook will have.

flamingbear replied:
I left the capitalization based on the keywords that are used in the actual request in the next cell.  I can change it though, or use fixed font.

flamingbear replied:
I like it fixed font

flamingbear replied:
151e797

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+    }
+   ],
+   "source": [
+    "import cartopy.crs as ccrs\n",
```
(Ignore if you want to): If you are using `cartopy` you could make the plot a tiny bit prettier, instead of just having the coastlines, you could do something like:

```
import cartopy.crs as ccrs
import cartopy.feature as cfeature

fig = plt.figure(figsize=(10,10))
fig, ax = plt.subplots(subplot_kw={'projection': ccrs.epsg(6933)})
ax.set_extent([x_coord_min, x_coord_max, y_coord_min, y_coord_max], crs=ccrs.epsg(6933))

# Feel free to use/not-use any of the below, but the land/ocean seem pretty
ax.add_feature(cfeature.LAND)
ax.add_feature(cfeature.OCEAN)
ax.add_feature(cfeature.COASTLINE)
ax.add_feature(cfeature.BORDERS, linestyle=':')
```

Disclaimer, I just ripped the `ax.add_feature` stuff from a different notebook and have not tested it in this case.

owenlittlejohns replied:
![Screenshot 2025-02-28 at 2 19 35 PM](https://github.com/user-attachments/assets/d0b68f32-cc18-47df-8ba0-23436e6e7ebd)

Here's what I got with that snippet (also commenting out `ax.coastlines()`), just in case you want to see a picture to judge if it really is any prettier.

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+   "source": [
+    "### Prerequisites\n",
+    "\n",
+    "The dependencies for this notebook are listed in the `docs-requirements.txt` file, launch your jupyter notebook from an environment that includes those dependencies.\n",
```
(You could ignore this if you like)

It might be worth just showing the `pip install -r doc-requirements.txt` command. (Similar thoughts to other places: I'm looking at this notebook trying to minimise how much we assume the reader knows)

flamingbear replied:
I'm kinda tired of holding every hand. I did have that actually originally but it seems like overkill. If you're looking to use this service, you should know enough python to create an environment and install the dependencies.  But if you feel strongly I can add it

flamingbear replied:
I put it in anyway.

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+    "\n",
+    "data_vars = in_datatree['Soil_Moisture_Retrieval_Data'].data_vars\n",
+    "print('First 10 data variables:')\n",
+    "list(data_vars)[0:10]"
```
More a question than a request for change - why did you do this instead of just using the `DataTree` HTML display? Was it because the attributes are automatically expanded and they have a whole bunch in them?

flamingbear replied:
because there's 53 variables in the repl and the attributes are annoying

flamingbear replied:
Well I may have changed my mind on that.

owenlittlejohns commented on docs/docs-requirements.txt:
```diff
@@ -0,0 +1,6 @@
+netCDF4==1.7.2
+numpy==2.2.3
+matplotlib==3.10.0
+Cartopy==0.24.1
```
Your requirements are missing `harmony-py`.

flamingbear replied:
0cfe43a

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+   "id": "f609c11a-173c-4a23-aab9-57bb38eab134",
+   "metadata": {},
+   "source": [
+    "### Locating grid values\n",
```
I like the explanation here showing what the service is doing with clear examples. Nice.

owenlittlejohns commented on docs/example-usage.ipynb:
```diff
@@ -0,0 +1,1478 @@
+    }
+   ],
+   "source": [
+    "job_id = harmony_client.submit(test_request)\n",
+    "\n",
+    "for filename in [\n",
+    "    file_future.result()\n",
+    "    for file_future in harmony_client.download_all(job_id, overwrite=True)\n",
```
Probably just an opinion...

I think what you are doing here is the most efficient in terms of lines of code (as `harmony.Client.download_all` will wait for processing to complete), but I think it might be a bit more user friendly if you used the `harmony.Client.wait_for_processing` method before, so you could use `show_progress=True`. It's mainly an aesthetic thing, but I think it's clearer/nicer for a user to see something saying Harmony is working on the request.

owenlittlejohns reviewed (changes requested):
This is a nice notebook, and I like the explanations of what the service actually does.

The one thing that needs fixing up is the missing `harmony-py` dependency from the requirements file. Everything else is opinion-based.

One other opinion - did you want a quick cell at the end saying something about how you can just use the services as a Python package, instead of a Harmony service? That feels valuable.

owenlittlejohns commented:
Oh, I didn't say in my review, but I executed the notebook, and once I'd installed the missing `harmony-py` dependency, everything worked as it should. Nice.

flamingbear committed:
DAS-2295: Update reqs for harmony-py

flamingbear committed:
DAS-2295: respond to PR feedback.

flamingbear committed:
DAS-2295: Last tweaks for notebook?

flamingbear commented:
@owenlittlejohns This is ready for re-review. Thanks for the previous comments

flamingbear committed:
DAS-2295: Clear the Datatree cells

These are interactive html cells that don't show any information in the
rendered notebook, but have a lot of useless data in them.

owenlittlejohns reviewed (approved):
Thanks for making all those tweaks - looks great!
